File: src/main/java/com/github/masooh/intellij/plugin/groovyfier/JUnitToSpockApplier.java
package com.github.masooh.intellij.plugin.groovyfier;

import static com.github.masooh.intellij.plugin.groovyfier.PsiHelper.changeMethodNameTo;
import static com.github.masooh.intellij.plugin.groovyfier.PsiHelper.removeStaticModifier;
import static com.github.masooh.intellij.plugin.groovyfier.PsiHelper.replaceElement;
import static com.github.masooh.intellij.plugin.groovyfier.PsiHelper.voidReturnToDef;

import java.util.Arrays;
import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.stream.Collectors;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.plugins.groovy.codeInspection.GroovyQuickFixFactory;
import org.jetbrains.plugins.groovy.lang.psi.GroovyPsiElementFactory;
import org.jetbrains.plugins.groovy.lang.psi.api.statements.GrLabeledStatement;
import org.jetbrains.plugins.groovy.lang.psi.api.statements.GrStatement;
import org.jetbrains.plugins.groovy.lang.psi.api.statements.expressions.GrBinaryExpression;
import org.jetbrains.plugins.groovy.lang.psi.api.statements.expressions.GrExpression;
import org.jetbrains.plugins.groovy.lang.psi.api.statements.expressions.GrUnaryExpression;
import org.jetbrains.plugins.groovy.lang.psi.api.statements.expressions.path.GrMethodCallExpression;
import org.jetbrains.plugins.groovy.lang.psi.api.statements.typedef.GrExtendsClause;
import org.jetbrains.plugins.groovy.lang.psi.api.statements.typedef.GrTypeDefinition;
import org.jetbrains.plugins.groovy.lang.psi.api.statements.typedef.members.GrMethod;
import org.jetbrains.plugins.groovy.lang.psi.impl.GroovyNamesUtil;
import org.jetbrains.plugins.groovy.lang.psi.impl.PsiImplUtil;

import com.intellij.codeInsight.intention.IntentionAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.PlatformDataKeys;
import com.intellij.openapi.command.WriteCommandAction;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.psi.PsiAnnotation;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiNameValuePair;
import com.intellij.psi.codeStyle.JavaCodeStyleManager;

public class JUnitToSpockApplier {
    private static final Logger LOG = Logger.getInstance(JUnitToSpockApplier.class);
    private final Project project;
    private final PsiFile psiFile;
    private final Editor editor;
    private final GrTypeDefinition typeDefinition;

    public JUnitToSpockApplier(AnActionEvent event) {
        project = event.getProject();
        psiFile = event.getRequiredData(PlatformDataKeys.PSI_FILE);
        editor = event.getRequiredData(PlatformDataKeys.EDITOR);
        typeDefinition = (GrTypeDefinition) PsiHelper.getPsiClass(psiFile);
    }

    void transformToSpock() {
        extendSpecification();
        changeMethods();

        // TODO add spock to dependencies if not present
        // TODO falls Test src/test/groovy anlegt bricht Umwandlung um: PSI and index do not match

        // TODO features durchgehen: https://github.com/opaluchlukasz/junit2spock
        // TODO Plugin/Feature fähig machen, wie bei junit2spock

        /* TODO Wicket feature: tester.assertComponent(FILE_LIST_PATH, ListView.class) -> then: (nur tester.assertX)
            nicht tester.lastRenderedPage.add(checkEventBehavior) oder tester.executeAjaxEvent(NEXT_STEP_LINK_PATH, "onclick")
         */
        // TODO first line then: -> expect:

        optimizeImports();
    }

    private void optimizeImports() {
        /* führt optimize nicht durch. Ursache unklar
            WriteCommandAction.runWriteCommandAction(project, () ->
                JavaCodeStyleManager.getInstance(project).optimizeImports(psiFile));
        */

        // TODO unklar was onTheFly ist
        IntentionAction fix = GroovyQuickFixFactory.getInstance().createOptimizeImportsFix(false);
        if (fix.isAvailable(project, editor, psiFile) && psiFile.isWritable()) {
            fix.invoke(project, editor, psiFile);
        }
    }

    private void changeMethods() {
        for (GrMethod method : typeDefinition.getCodeMethods()) {

            changeMethodHavingAnnotation(method, "org.junit.Test", () -> {
                changeMethodNameTo(method, "\"" + camelToSpace(method.getName()) + "\"");
                voidReturnToDef(method);
                changeMethodBody(method);
            });

            // TODO handle JUnit 5 annotations
            changeMethodHavingAnnotation(method, "org.junit.Before", () -> {
                changeMethodNameTo(method, "setup");
                voidReturnToDef(method);
            });

            changeMethodHavingAnnotation(method, "org.junit.After", () -> {
                changeMethodNameTo(method, "cleanup");
                voidReturnToDef(method);
            });

            changeMethodHavingAnnotation(method, "org.junit.BeforeClass", () -> {
                changeMethodNameTo(method, "setupSpec");
                removeStaticModifier(method);
                voidReturnToDef(method);
            });

            changeMethodHavingAnnotation(method, "org.junit.AfterClass", () -> {
                changeMethodNameTo(method, "cleanupSpec");
                removeStaticModifier(method);
                voidReturnToDef(method);
            });
        }
    }

    /**
     * also deletes the annotation
     */
    private void changeMethodHavingAnnotation(GrMethod method, String annotationName,
                                              Runnable changeInMethod) {

        PsiAnnotation annotation = PsiImplUtil.getAnnotation(method, annotationName);

        if (annotation != null) {
            Object exceptionClass = null;
            for (PsiNameValuePair attribute : annotation.getParameterList().getAttributes()) {
                switch (attribute.getName()) {
                    case "expected":
                        exceptionClass = attribute.getValue();
                        break;
                    // TODO timeout
                    default:
                        LOG.error("unhandled attribute: {}", attribute.getName());
                }
            }

            WriteCommandAction.runWriteCommandAction(project, annotation::delete);
            WriteCommandAction.runWriteCommandAction(project, changeInMethod);
        }
    }

    private void changeMethodBody(GrMethod method) {
        addWhenToFirstStatement(method);
        replaceAsserts(method);
    }

    private void replaceAsserts(GrMethod method) {
        List<GrStatement> methodCalls = Arrays.stream(method.getBlock().getStatements())
                .filter(grStatement -> grStatement instanceof GrMethodCallExpression)
                .collect(Collectors.toList());

        AtomicBoolean firstAssertion = new AtomicBoolean(true);

        methodCalls.stream()
                .filter(grStatement -> grStatement instanceof GrMethodCallExpression)
                .map(grStatement -> (GrMethodCallExpression)grStatement)
                .filter(methodCallExpression -> {
                    String text = methodCallExpression.getFirstChild().getText();
                    // with or without import
                    return text.startsWith("assert") || text.startsWith("Assert.");
                }).forEach(methodCall -> {
            GrExpression spockAssert = getSpockAssert(methodCall);

            if (firstAssertion.getAndSet(false)) {
                GrStatement spockAssertWithLabel = getFactory().createStatementFromText("then: expression");
                GrExpression grExpression = (GrExpression) spockAssertWithLabel.getLastChild();
                grExpression.replaceWithExpression(spockAssert, true);
                replaceElement(methodCall, spockAssertWithLabel);
            } else {
                replaceElement(methodCall, spockAssert);
            }
        });
    }

    private GrExpression getSpockAssert(GrMethodCallExpression methodCallExpression) {
        GrExpression[] expressionArguments = methodCallExpression.getArgumentList().getExpressionArguments();

        GrExpression firstArgument = expressionArguments[0];
        GrExpression secondArgument = expressionArguments.length > 1 ? expressionArguments[1] : null;

        GrExpression spockAssert;

        // remove Assert class if there
        String methodName = methodCallExpression.getFirstChild().getText().replace("Assert.", "");
        switch (methodName) {
            case "assertEquals":
                GrBinaryExpression equalsExpression = createExpression("actual == expected");
                equalsExpression.getLeftOperand().replaceWithExpression(secondArgument, true);
                equalsExpression.getRightOperand().replaceWithExpression(firstArgument, true);
                spockAssert = equalsExpression;
                break;
            case "assertTrue":
                spockAssert = createExpression("actual").replaceWithExpression(firstArgument, true);
                break;
            case "assertFalse":
                GrUnaryExpression unaryExpression = createExpression("!actual");
                unaryExpression.getOperand().replaceWithExpression(firstArgument, true);
                spockAssert = unaryExpression;
                break;
            case "assertNotNull":
                GrBinaryExpression notNullExpression = createExpression("actual != null");
                notNullExpression.getLeftOperand().replaceWithExpression(firstArgument, true);
                spockAssert = notNullExpression;
                break;
            case "assertNull":
                GrBinaryExpression nullExpression = createExpression("actual == null");
                nullExpression.getLeftOperand().replaceWithExpression(firstArgument, true);
                spockAssert = nullExpression;
                break;
            default:
                throw new IllegalArgumentException("Unknown assert " + methodName);
        }
        return spockAssert;
    }

    private void addWhenToFirstStatement(GrMethod grMethod) {
        GrStatement firstStatement = grMethod.getBlock().getStatements()[0];

        GrLabeledStatement firstStatementWithWhen = createStatementFromText("when: expression");
        firstStatementWithWhen.getStatement().replaceWithStatement(firstStatement);

        replaceElement(firstStatement, firstStatementWithWhen);
    }

    @NotNull
    private <T extends GrStatement> T createStatementFromText(String expression) {
        return (T) getFactory().createStatementFromText(expression);
    }

    @NotNull
    private <T extends GrExpression> T createExpression(String expression) {
        return (T) getFactory().createExpressionFromText(expression);
    }

    /**
     * only insert extends if no extends are there
     */
    private void extendSpecification() {
        if (typeDefinition.getExtendsList().getTextLength() == 0) {
            GrTypeDefinition definition = getFactory().createTypeDefinition("class A extends spock.lang.Specification {}");
            GrExtendsClause extendsClause = definition.getExtendsClause();

            // ask welchen Effekt hat command und groupID?
            WriteCommandAction.runWriteCommandAction(project, null, null, () -> {
                PsiElement addedExtend = typeDefinition.addAfter(extendsClause, typeDefinition.getNameIdentifierGroovy());

                JavaCodeStyleManager.getInstance(project).shortenClassReferences(addedExtend);
            }, psiFile);
        }
    }

    @NotNull
    private GroovyPsiElementFactory getFactory() {
        return GroovyPsiElementFactory.getInstance(project);
    }


    private String camelToSpace(final String string) {
        return StringUtil.join(GroovyNamesUtil.camelizeString(string), StringUtil::decapitalize, " ");
    }
}

File: src/main/kotlin/com/github/masooh/intellij/plugin/groovyfier/JUnitToSpockApplier.kt
package com.github.masooh.intellij.plugin.groovyfier

import com.github.masooh.intellij.plugin.groovyfier.Block.*
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.PlatformDataKeys
import com.intellij.openapi.command.WriteCommandAction
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.text.StringUtil
import com.intellij.psi.JavaPsiFacade
import com.intellij.psi.PsiAnnotation
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.codeStyle.JavaCodeStyleManager
import com.intellij.psi.impl.source.tree.PsiWhiteSpaceImpl
import org.jetbrains.plugins.groovy.codeInspection.GroovyQuickFixFactory
import org.jetbrains.plugins.groovy.lang.psi.GroovyPsiElementFactory
import org.jetbrains.plugins.groovy.lang.psi.api.statements.GrLabeledStatement
import org.jetbrains.plugins.groovy.lang.psi.api.statements.GrStatement
import org.jetbrains.plugins.groovy.lang.psi.api.statements.GrVariableDeclaration
import org.jetbrains.plugins.groovy.lang.psi.api.statements.expressions.GrBinaryExpression
import org.jetbrains.plugins.groovy.lang.psi.api.statements.expressions.GrExpression
import org.jetbrains.plugins.groovy.lang.psi.api.statements.expressions.GrReferenceExpression
import org.jetbrains.plugins.groovy.lang.psi.api.statements.expressions.GrUnaryExpression
import org.jetbrains.plugins.groovy.lang.psi.api.statements.expressions.path.GrMethodCallExpression
import org.jetbrains.plugins.groovy.lang.psi.api.statements.typedef.GrTypeDefinition
import org.jetbrains.plugins.groovy.lang.psi.api.statements.typedef.members.GrMethod
import org.jetbrains.plugins.groovy.lang.psi.impl.GroovyNamesUtil
import org.jetbrains.plugins.groovy.lang.psi.impl.PsiImplUtil

enum class Block {
    EXPECT, GIVEN, WHEN, THEN;

    val label
        get() = this.name.toLowerCase()
}

class JUnitToSpockApplier(event: AnActionEvent) {
    companion object {
        private val log = Logger.getInstance(JUnitToSpockApplier::class.java)
    }

    private val project: Project = event.project!!
    private val psiFile: PsiFile = event.getRequiredData(PlatformDataKeys.PSI_FILE)
    private val editor: Editor = event.getRequiredData(PlatformDataKeys.EDITOR)
    private val typeDefinition: GrTypeDefinition

    private val groovyFactory
        get() = GroovyPsiElementFactory.getInstance(project)

    private val javaFactory
        get() = JavaPsiFacade.getInstance(project).elementFactory

    init {
        typeDefinition = psiFile.getPsiClass() as GrTypeDefinition
    }

    fun transformToSpock() {
        // spock has it's own runner
        typeDefinition.getAnnotation("org.junit.runner.RunWith")?.delete()

        extendSpecification()
        changeMethods()

        // TODO add spock to dependencies if not present
        // TODO falls Test src/test/groovy anlegt bricht Umwandlung um: PSI and index do not match

        // TODO features durchgehen: https://github.com/opaluchlukasz/junit2spock
        // TODO Plugin/Feature fähig machen, wie bei junit2spock

        /* TODO Wicket feature: tester.assertComponent(FILE_LIST_PATH, ListView.class) -> then: (nur tester.assertX)
            nicht tester.lastRenderedPage.add(checkEventBehavior) oder tester.executeAjaxEvent(NEXT_STEP_LINK_PATH, "onclick")
         */
        optimizeImports()
    }

    private fun optimizeImports() {
        /* führt optimize nicht durch. Ursache unklar
            WriteCommandAction.runWriteCommandAction(project, () ->
                JavaCodeStyleManager.getInstance(project).optimizeImports(psiFile));
        */

        // TODO unklar was onTheFly ist
        val fix = GroovyQuickFixFactory.getInstance().createOptimizeImportsFix(false)
        if (fix.isAvailable(project, editor, psiFile) && psiFile.isWritable) {
            fix.invoke(project, editor, psiFile)
        }
    }

    private fun changeMethods() {
        for (method in typeDefinition.codeMethods) {

            // TODO extract changes in list and apply only one of it
            changeMethodHavingAnnotation(method, "org.junit.Test", "org.junit.jupiter.api.Test") { annotation ->
                var exceptionClass: GrReferenceExpression? = null
                for (attribute in annotation.parameterList.attributes) {
                    when (attribute.name) {
                        "expected" -> exceptionClass = attribute.value as GrReferenceExpression
                        // TODO timeout
                        else -> log.error("unhandled attribute: {}", attribute.name)
                    }
                }

                method.changeMethodNameTo("\"" + camelToSpace(method.name) + "\"")
                        .voidReturnToDef()

                changeFeatureBody(method)

                if (exceptionClass != null) {
                    val statement = "then: thrown(${exceptionClass.qualifierExpression!!.text})"
                    val thrownBlock = createStatementFromText<GrLabeledStatement>(statement)

                    // add would insert statement after closing }
                    method.block!!.addBefore(thrownBlock, method.block!!.lastChild)
                }
            }

            changeMethodHavingAnnotation(method, "org.junit.Before", "org.junit.jupiter.api.BeforeEach") {
                method.changeMethodNameTo("setup")
            }

            changeMethodHavingAnnotation(method, "org.junit.After", "org.junit.jupiter.api.AfterEach") {
                method.changeMethodNameTo("cleanup")
            }

            changeMethodHavingAnnotation(method, "org.junit.BeforeClass", "org.junit.jupiter.api.BeforeAll") {
                method.changeMethodNameTo("setupSpec")
                        .removeStaticModifier()
            }

            changeMethodHavingAnnotation(method, "org.junit.AfterClass", "org.junit.jupiter.api.AfterAll") {
                method.changeMethodNameTo("cleanupSpec")
                        .removeStaticModifier()
            }
        }
    }

    /**
     * also deletes the annotation and replaces void with def
     */
    private fun changeMethodHavingAnnotation(method: GrMethod, vararg annotationName: String,
                                             changeInMethod: (PsiAnnotation) -> Unit) {


        val annotation = annotationName.asSequence().mapNotNull { PsiImplUtil.getAnnotation(method, it) }.firstOrNull()

        if (annotation != null) {
            // todo macht es Sinn für jede Methode extra Write action?
            WriteCommandAction.runWriteCommandAction(project) {
                changeInMethod(annotation)
                annotation.delete()
                method.voidReturnToDef()
            }
        }
    }

    private fun changeFeatureBody(method: GrMethod) {
        val statements = method.block?.statements

        var currentBlock: Block? = null

        statements?.forEach { statement ->
            when (currentBlock) {
                null -> {
                    log.info("null:")
                    currentBlock = when {
                        statement.isAssertion() -> {
                            val replacedStatement = replaceWithSpockAssert(statement as GrMethodCallExpression)
                            addLabelToStatement(EXPECT, replacedStatement)
                        }
                        else -> {
                            addLabelToStatement(WHEN, statement)
                        }
                    }
                }
                WHEN -> {
                    log.info("when:")
                    if (statement.isAssertion()) {
                        // todo hier wird ersetzt, um das ersetzte wieder zu ersetzen
                        //    kann man das in einem machen? statement replace with (label + spock assert)
                        val statementWithSpockAssertion = replaceWithSpockAssert(statement as GrMethodCallExpression)
                        currentBlock = addLabelToStatement(THEN, statementWithSpockAssertion)
                    }
                }
                EXPECT -> {
                    log.info("expect:")
                    val block = handleExpectAndThen(statement)
                    if (block != null) {
                        currentBlock = block
                    }
                }
                GIVEN -> {
                    log.info("given:")
                    TODO("given not implemented yet")
                }
                THEN -> {
                    log.info("then:")
                    val block = handleExpectAndThen(statement)
                    if (block != null) {
                        currentBlock = block
                    }
                }
            }

        }
    }

    private fun handleExpectAndThen(statement: GrStatement): Block? {
        return when {
            statement.isAssertion() -> {
                replaceWithSpockAssert(statement as GrMethodCallExpression)
                null // stay in block
            }
            statement is GrVariableDeclaration -> {
                null // stay in block
            }
            else -> {
                addLabelToStatement(WHEN, statement) // next when
            }
        }
    }

    private fun addLabelToStatement(block: Block, statement: GrStatement): Block {
        val statementWithLabel = groovyFactory.createStatementFromText("${block.label}: statement", statement.parent)
        val textStatement = statementWithLabel.lastChild as GrStatement
        textStatement.replaceWithStatement(statement)

        // todo https://youtrack.jetbrains.com/issue/IDEA-185879
        //  wie komme ich an attachement

        statement.replaceWithStatement(statementWithLabel)
        return block
    }

    private fun replaceWithSpockAssert(methodCallExpression: GrMethodCallExpression): GrExpression {
        val argumentList = methodCallExpression.argumentList

        var spockAssert: GrExpression? = null
        var message: GrExpression? = null

        // remove Assert class if there
        val methodName = methodCallExpression.firstChild.text.replace("Assert.", "")

        when (methodName) {
            "assertEquals" -> {
                spockAssert = argumentList.withArgs(
                        two = { expected, actual ->
                            createBinaryExpression("actual == expected", actual, expected)
                        },
                        three = { msg, expected, actual ->
                            message = msg
                            createBinaryExpression("actual == expected", actual, expected)
                        })
            }
            "assertNotEquals" -> {
                spockAssert = argumentList.withArgs(
                        two = { expected, actual ->
                            createBinaryExpression("actual != unexpected", actual, expected)
                        },
                        three = { msg, expected, actual ->
                            message = msg
                            createBinaryExpression("actual != unexpected", actual, expected)
                        })
            }
            "assertTrue" -> {
                argumentList.withArgs(
                        one = { it },
                        two = { msg, cond ->
                            message = msg
                            cond
                        })?.let {
                    spockAssert = createExpression("condition").replaceWithExpression(it, true)
                }
            }
            "assertFalse" -> {
                argumentList.withArgs(
                        one = { it },
                        two = { msg, condition ->
                            message = msg
                            condition
                        })?.let {
                    val unaryExpression = createExpression<GrUnaryExpression>("!actual")
                    unaryExpression.operand!!.replaceWithExpression(it, true)
                    spockAssert = unaryExpression
                }
            }
            "assertNotNull" -> {
                argumentList.withArgs(
                        one = { it },
                        two = { msg, obj ->
                            message = msg
                            obj
                        }
                )?.let {
                    spockAssert = createExpression<GrBinaryExpression>("actual != null").apply {
                        leftOperand.replaceWithExpression(it, true)
                    }
                }
            }
            "assertNull" -> {
                argumentList.withArgs(
                        one = { it },
                        two = { msg, obj ->
                            message = msg
                            obj
                        })?.let {
                    val nullExpression = createExpression<GrBinaryExpression>("actual == null")
                    nullExpression.leftOperand.replaceWithExpression(it, true)
                    spockAssert = nullExpression
                }
            }
            else -> {
                log.warn("Unknown assert $methodName")
            }
        }

        log.info("Replace $methodName, args: ${argumentList.expressionArguments.size} with $spockAssert")

        return spockAssert?.let { assertion ->
            // actual replacement if not done above
            val replacedExpression = methodCallExpression.replaceWithExpression(assertion, true)

            message?.let {
                // add message as comment
                val whiteSpaceAndComment = replacedExpression.createComment(it.text)
                replacedExpression.addRangeAfter(whiteSpaceAndComment)
            }
            replacedExpression
        } ?: methodCallExpression
    }

    private fun createBinaryExpression(expression: String, left: GrExpression, right: GrExpression): GrBinaryExpression {
        val equalsExpression = createExpression<GrBinaryExpression>(expression)
        equalsExpression.leftOperand.replaceWithExpression(left, true)
        equalsExpression.rightOperand!!.replaceWithExpression(right, true)
        return equalsExpression
    }

    @Suppress("UNCHECKED_CAST")
    private fun <T : GrStatement> createStatementFromText(expression: String): T {
        return groovyFactory.createStatementFromText(expression) as T
    }

    @Suppress("UNCHECKED_CAST")
    private fun <T : GrExpression> createExpression(expression: String): T {
        return groovyFactory.createExpressionFromText(expression) as T
    }

    /**
     * only insert extends if no extends are there
     */
    private fun extendSpecification() {
        if (typeDefinition.extendsList?.textLength == 0) {
            val definition = groovyFactory.createTypeDefinition("class A extends spock.lang.Specification {}")
            val extendsClause = definition.extendsClause!!

            // ask welchen Effekt hat command und groupID?
            WriteCommandAction.runWriteCommandAction(project, null, null, Runnable {
                val addedExtend = typeDefinition.addAfter(extendsClause, typeDefinition.nameIdentifierGroovy)
                JavaCodeStyleManager.getInstance(project).shortenClassReferences(addedExtend)
            }, psiFile)
        }
    }


    private fun camelToSpace(string: String): String {
        return StringUtil.join(GroovyNamesUtil.camelizeString(string), { StringUtil.decapitalize(it) }, " ")
    }
}

private fun GrStatement.isAssertion(): Boolean {
    return if (this is GrMethodCallExpression) {
        val text = this.firstChild.text
        // with or without import
        text.startsWith("assert") || text.startsWith("Assert.")
    } else {
        false
    }
}
